security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,412 @@
+          this._tpm.withSession(tss.NullPwSession).HMAC(TpmSecurityClient._idKeyPersistentHandle, dataToSign, idKeyHashAlg, (signature: Buffer) => {
+            callback(null, signature);
+          });
+        } else {
+          let curPos: number = 0;
+          let bytesLeft: number = dataToSign.length;
+          let hSequence: TPM_HANDLE = null;
+          let signature = new Buffer(0);
+          let loopFn = () => {
+            if (bytesLeft > maxInputBuffer) {
+                this._tpm.withSession(tss.NullPwSession).SequenceUpdate(hSequence, dataToSign.slice(curPos, curPos + maxInputBuffer), loopFn);
+                console.log('SequenceUpdate() invoked for slice [' + curPos + ', ' + (curPos + maxInputBuffer) + ']');
+                bytesLeft -= maxInputBuffer;
+                curPos += maxInputBuffer;
+            } else {
+              this._tpm.withSession(tss.NullPwSession).SequenceComplete(hSequence, dataToSign.slice(curPos, curPos + bytesLeft), new TPM_HANDLE(tss.TPM_RH.NULL), (resp: tss.SequenceCompleteResponse) => {
+                console.log('SequenceComplete() succeeded; signature size ' + signature.length);
+              });
+            }
+          };
+          this._tpm.withSession(tss.NullPwSession).HMAC_Start(TpmSecurityClient._idKeyPersistentHandle, signature, idKeyHashAlg, (hSeq: TPM_HANDLE) => {
+            console.log('HMAC_Start() returned ' + TPM_RC[this._tpm.getLastResponseCode()]);
+            hSequence = hSeq;
+            loopFn();
+          });
+        }
+      }
+    });
   }
 
-  activateSymmetricIdentity(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  private _activateSymmetricIdentity(activationBlob: Buffer, activateCallback: (err: Error) => void): void {
+
+    let currentPosition = 0;
+    let credentialBlob: tss.TPMS_ID_OBJECT;
+    let encodedSecret = new tss.TPM2B_ENCRYPTED_SECRET();
+    let idKeyDupBlob = new TPM2B_PRIVATE();
+    let encWrapKey = new tss.TPM2B_ENCRYPTED_SECRET();
+
+    //
+    // Unmarshal components of the activation blob received from the provisioning service.
+    //
+    [credentialBlob, currentPosition] = tss.marshal.sizedFromTpm(tss.TPMS_ID_OBJECT, activationBlob, 2, currentPosition);
+    debug('credentialBlob end: ' + currentPosition);
+    currentPosition = encodedSecret.fromTpm(activationBlob, currentPosition);
+    debug('encodedSecret end: ' + currentPosition);
+    currentPosition = idKeyDupBlob.fromTpm(activationBlob, currentPosition);
+    debug('idKeyDupBlob end: ' + currentPosition);
+    currentPosition = encWrapKey.fromTpm(activationBlob, currentPosition);
+    debug('encWrapKey end: ' + currentPosition);
+    [this._idKeyPub, currentPosition] = tss.marshal.sizedFromTpm(TPMT_PUBLIC, activationBlob, 2, currentPosition);
+    debug('idKeyPub end: ' + currentPosition);
+
+    //
+    // Start a policy session to be used with ActivateCredential()
+    //
+
+    this._tpm.GetRandom(20, (nonce: Buffer) => {
```

Comment:
>20 [](start = 24, length = 2)

why?

Reply:
Because that's the size of the nonce needed to start an auth session.  Have provided a reference and made it a static readonly value.

---
In reply to: [160301911](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r160301911) [](ancestors = 160301911)